pkg/node/address_linux.go

```diff
@@ -97,10 +107,18 @@ retryScope:
 		// Just make sure that we always return the same one and not a
 		// random one. More info in the issue GH-7637.
 		sort.Slice(ipsPublic, func(i, j int) bool {
-			return bytes.Compare(ipsPublic[i], ipsPublic[j]) < 0
+			if ipsPublic[i].IsSecondary() != ipsPublic[j].IsSecondary() {
+				switch {
+				case ipsPublic[i].IsSecondary():
+					return false
+				case ipsPublic[j].IsSecondary():
+					return true
+				}
+			}
+			return bytes.Compare(ipsPublic[i].IP, ipsPublic[j].IP) < 0
 		})
```

Comment:
Maybe we can move this into a function to avoid duplicating it below:
```go
sortIPs := func(i, j int) bool {
	return bytes.Compare(ipsPublic[i], ipsPublic[j]) < 0
	if ipsPublic[i].IsSecondary() != ipsPublic[j].IsSecondary() {
		switch {
		case ipsPublic[i].IsSecondary():
			return false
		case ipsPublic[j].IsSecondary():
			return true
		}
	}
	return bytes.Compare(ipsPublic[i].IP, ipsPublic[j].IP) < 0
}

Reply:
Nice drive-by, but sadly it's not that easy since we need to sort two different slice.
I'll add it, but it'll be slightly less readable due to the passing of a slice by pointer.